README.md
# node-timecapsule

[![Greenkeeper badge](https://badges.greenkeeper.io/gabrielcsapo/node-timecapsule.svg)](https://greenkeeper.io/)

[![Npm Version](https://img.shields.io/npm/v/node-timecapsule.svg)](https://www.npmjs.com/package/node-timecapsule)
[![Dependency Status](https://david-dm.org/gabrielcsapo/node-timecapsule.svg)](https://david-dm.org/gabrielcsapo/node-timecapsule)
[![devDependency Status](https://david-dm.org/gabrielcsapo/node-timecapsule/dev-status.svg)](https://david-dm.org/gabrielcsapo/node-timecapsule#info=devDependencies)
[![Build Status](https://travis-ci.org/gabrielcsapo/node-timecapsule.svg?branch=master)](https://travis-ci.org/gabrielcsapo/node-timecapsule)
[![npm](https://img.shields.io/npm/dt/node-timecapsule.svg)]()
[![npm](https://img.shields.io/npm/dm/node-timecapsule.svg)]()

> saving the internet for future generations

# Installation

```
npm install node-timecapsule
```

# Usage

> programmatic

```javascript
const Timecapsule = require('node-timecapsule');
const timecapsule = new Timecapsule(); // you can pass a directory to the constructor if you want to specific a different directory then the default, `capsules`.

timecapsule.save('http://www.bestbuy.com', (error, result) = > {
    // will return the resources saved or an error
});
```

> shell

```
$ timecapsule save http://www.bestbuy.com
```

# Screenshots

![main](./screenshots/main.png)

![backup](./screenshots/backup.png)


CHANGELOG.md
# 0.2.0

- save, get and getAll all require a callback.
- UI has been further simplified

# 0.1.0

- changes the modules interface and adds front-end component.
- adds option to be used as a command line tool

# 0.0.1

- very basic web scraper, needs work on maintaining structure and viewing saved capsules

# backlog

- [x] figure out why `Segmentation fault: 11` keep happening in `website-scraper`
- [x] place capsules in the correct directory (timestamp them?)
- [x] be able to view saved capsules?
    - might just be a page that can search all of the saved files?
- write tests
- might be interesting instead of saving all of the html to just over an option to take a screenshot of all the pages?


test.js
var timecapsule = require('./index');

timecapsule.capsulate('http://www.bestbuy.com');


util.js
var url = require('url');
var scraper = require('website-scraper');
var Horseman = require('node-horseman');
var horseman = new Horseman();

module.exports = {
    save: function(urls, callback) {
        var options = {
            urls: urls,
            prettifyUrls: true,
            directory: './capsules',
            recursive: true,
            maxDepth: 10,
            request: {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0'
                }
            }
        };

        scraper.scrape(options, function(error, result) {
            callback(result, error);
        });
    },
    /**
     * finds all a tags on page to scrape
     * @function scrapeUrls
     * @property uri - website url (necessary)
     * @property callback - function(value, error)
     */
    scrapeUrls: function(uri, callback) {
        // This code gives us all the paths of the links on the uri we requested
        horseman
            .userAgent('Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0')
            .open(uri)
            .screenshot('/capsules/' + encodeURIComponent(escape(uri)) + '.png')
            .evaluate(function() {
                return [].map.call(document.querySelectorAll('a'), function(link) {
                    return link.getAttribute('href');
                });
            })
            .then(function(value) {
                value = value.filter(function(u) {
                    // remove all urls that do not work
                    return url.parse(u).path && url.parse(u).path.indexOf('void(0)') == -1;
                }).map(function(u) {
                    // if the urls do not have a hostname
                    // append the url because they are relative
                    if (url.parse(u).path && !url.parse(u).hostname) {
                        return uri + u;
                    } else {
                        return u;
                    }
                });
                callback(value, null);
            })
            .catch(function(e) {
                callback(null, e);
            })
            .close()
    }
}


src/main.js
import 'whatwg-fetch';

import React from 'react';

class Main extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            result: [],
            loading: false
        };
    }
    componentWillMount() {
        this.setState({
            loading: true
        });

        fetch(`/api/a`)
            .then((response) => {
                return response.json();
            })
            .then((json) => {
                const { result } = json;
                this.setState({
                    result,
                    loading: false
                });
            })
            .catch(() => {
                this.setState({
                    loading: false
                });
            });
    }
    render () {
        const { loading, result } = this.state;

        return (
            <div className="text-center" style={{ width:"50%", position: "absolute", top: "50%", transform: "translateY(-50%)", right: '25%'}}>
                { loading ?
                    <ul className="list" style={{ padding: '20px' }}>
                        <li className="list-item">
                            <div className="spinner-wrapper">
                                <div className="spinner spinner-primary"></div>
                            </div>
                        </li>
                    </ul>
                :
                    <div>
                        { result.length > 0 ?
                            <ul className="list" style={{ padding: '20px' }}>
                                { result.map((result, i) => {
                                    return (
                                        <li className="list-item" key={ i }>
                                            <a href={ `/backup/?searchTerm=${result.url}` } target="_blank">
                                                { result.url } - { result.dates.length } <small> { result.dates.length == 1 ? 'backup' : 'backups' }</small>
                                            </a>
                                        </li>
                                    );
                                })}
                            </ul>
                        :
                            <ul className="list" style={{ padding: '20px' }}>
                                <li className="list-item">
                                    no results found 🙈
                                </li>
                                <li className="list-item">
                                    <a href="/backup"> Back Something Up! </a>
                                </li>
                            </ul>
                        }
                    </div>
                }
            </div>
        );
    }
}

export default Main;


src/backup.js
import 'whatwg-fetch';

import React from 'react';
import Moment from 'moment';
import Url from 'url';
import QS from 'querystring';

class Search extends React.Component {
  constructor(props) {
      super(props);
      const url = Url.parse(location.href);
      const query = QS.parse(url.query);

      this.state = {
          result: {},
          searchTerm: query.searchTerm || '',
          request: 'done' // loading, done, error
      };
      this.get(this.state.searchTerm);
  }
  get(url) {
      this.setState({
          request: 'loading'
      });

      fetch(`/api/i/${url}`)
          .then((response) => {
              return response.json();
          })
          .then((json) => {
              const { result } = json;

              this.setState({
                  result,
                  searchTerm: url,
                  request: 'done'
              });
          })
          .catch(function(ex) {
              console.log('parsing failed', ex); // eslint-disable-line
          });
  }
  search(e) {
      const searchTerm = e.target.value;

      if(e.keyCode == 13) {
        this.get(searchTerm);
        if (history.pushState) {
            var newurl = window.location.protocol + "//" + window.location.host + window.location.pathname + '?searchTerm=' + searchTerm;
            window.history.pushState({path:newurl},'',newurl);
        }
      }
  }
  save() {
      const { searchTerm } = this.state;
      this.setState({
          request: 'loading'
      });
      fetch(`/api/b/${searchTerm}`)
          .then((response) => {
              return response.json();
          })
          .then((json) => {
              if(json.error) {
                  this.setState({
                      request: 'error'
                  });
              } else {
                  this.setState({
                      request: 'done'
                  });
                  this.get(searchTerm);
              }
          }).catch(function(ex) {
              console.log('parsing failed', ex); // eslint-disable-line
          });
  }
  render() {
    const { result, searchTerm, request } = this.state;
    let buttonValue = <button></button>;

    if(request == 'loading' && searchTerm) {
        buttonValue = <button style={{ margin: 0, height: "39px", padding: ".375rem 1% .375rem 1%" }} className="btn btn-block"><div className="spinner spinner-primary"></div></button>;
    } else if(request == 'done' && searchTerm) {
        buttonValue = <button style={{ margin: "0 5px", height: "39px", padding: ".375rem 1% .375rem 1%" }} onClick={this.save.bind(this)} className="btn btn-block"> Backup </button>;
    } else {
        buttonValue = <button style={{ margin: 0, height: "39px", padding: ".375rem 1% .375rem 1%" }} onClick={this.search.bind(this)} className="btn btn-block"> Search </button>;
    }

    return (
        <div style={{ width:"50%", margin: "0 auto" }}>
            <div style={{ padding: '20px', margin: '0 auto', width: '80%' }}>
                <div className="grid">
                    <div className="col-10-12">
                        <input style={{ height: "25px" }} onKeyDown={this.search.bind(this)} onKeyUp={this.search.bind(this)} type="text" placeholder={ searchTerm || "Enter the url you want view" }/>
                    </div>
                    <div className="col-2-12">
                        { buttonValue }
                    </div>
                </div>
                <br/>
                <div className="grid">
                    <div className="col-12-12">
                        { request === 'loading' ?
                            <ul className="list">
                                <div className="spinner-wrapper">
                                    <div className="spinner spinner-primary"></div>
                                </div>
                            </ul>
                        :
                        <div>
                            { Object.keys(result).length > 0 ?
                            <div>
                                <ul className="list">
                                    { result.dates.map((date, i) => { return (
                                    <li className="list-item" key={ i }>
                                        <a href={ `/v/${result.hash}/${date}/index.html` } target="_blank">
                                            { Moment(date, 'Y-M-D-h:mm:ss:a').format() } ({ Moment(date, 'Y-M-D-h:mm:ss:a').fromNow() })
                                        </a>
                                    </li>
                                    ); })}
                                </ul>
                            </div>
                            :
                            <div>
                                <ul className="list">
                                    <li className="list-item text-center">
                                        no results found 🙈
                                    </li>
                                </ul>
                            </div>
                            }
                        </div>
                        }
                    </div>
                </div>
            </div>
        </div>
    );
  }
}

export default Search;


src/layout.js
import React from 'react';
import PropTypes from 'prop-types';

class Layout extends React.Component {
  render () {
    const { children } = this.props;

    return (
      <div style={{ height: '100%', width: '100%' }}>
        <div className="navbar navbar-center">
          <div className="container">
            <div className="navbar-title">
              <a className="text-black" href="/">
                <span style={{ fontSize: '30px' }} className="text-black">node-timecapsule</span>
              </a>
            </div>
            <div className="nav">
                <a href="/">Main</a>
                <a href="/backup">Backup</a>
            </div>
          </div>
        </div>
        <div style={{ marginBottom: "60px" }}>
          { children }
        </div>
      </div>
    );
  }
}

Layout.propTypes = {
  children: PropTypes.object
};

export default Layout;


webpack.config.js
const webpack = require('webpack');
const path = require('path');

const config = {
    entry: './src/app.js',
    output: {
        path: path.resolve(__dirname, 'dist'),
        filename: 'build.js'
    },
    devServer: {
        port: 5000,
        contentBase: 'dist',
        proxy: {
            '/api/**': 'http://localhost:3000',
            '/v/**': 'http://localhost:3000'
        },
        inline: true,
        historyApiFallback: true
    },
    module: {
        loaders: [{
                test: /\.css$/,
                loaders: ['style-loader', 'css-loader']
            }, {
                test: /.jsx?$/,
                loader: 'babel-loader',
                exclude: /node_modules/,
                query: {
                    presets: ['es2015', 'react']
                }
            }
        ]
    }
};

if (process.env.NODE_ENV == 'production') {
    config.plugins = [
        new webpack.DefinePlugin({
            'process.env': {
                'NODE_ENV': JSON.stringify('production')
            }
        }),
        new webpack.ContextReplacementPlugin(/moment[\\]locale$/, /^\.\/(en)$/),
        new webpack.optimize.UglifyJsPlugin({
            comments: false,
            compress: {
                unused: true,
                dead_code: true,
                warnings: false,
                drop_debugger: true,
                conditionals: true,
                evaluate: true,
                sequences: true,
                booleans: true,
            }
        }),
        new webpack.optimize.AggressiveMergingPlugin(),
    ];
}

module.exports = config;


index.js
const scrape = require('website-scraper');
const moment = require('moment');
const fs = require('fs');
const path = require('path');
const express = require('express');
const async = require('async');
const compression = require('compression');
const normalizeUrl = require('normalize-url');

class Timecapsule {
    constructor(directory) {
        this.directory = directory || path.resolve(__dirname, 'capsules');
    }
    /**
     * this will save the url to a folder that is hashed to the url value and with sub folders with the dates they are saved
     * @function save
     * @property url - url of a website
     */
    save(url, callback) {
        url = normalizeUrl(url);

        const hashUrl = url.indexOf('://') > -1 ? url.substring(url.indexOf('://') + 3, url.length) : url;
        const hash = new Buffer(hashUrl).toString('base64');
        const date = moment().format('Y-M-D-h:mm:ss:a');

        const options = {
            urls: [url],
            prettifyUrls: true,
            recursive: true,
            maxRecursiveDepth: 2,
            maxDepth: 2,
            directory: `${this.directory}/${hash}/${date}`,
            request: {
                headers: {
                    'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:27.0) Gecko/20100101 Firefox/27.0'
                }
            }
        };

        scrape(options)
            .then((result) => {
                callback ? callback(null, result) : console.log('success!'); // eslint-disable-line
            })
            .catch((err) => {
                callback ? callback(err, null) : console.error(err); // eslint-disable-line
            });
    }
    get(url, options, callback) {
        url = normalizeUrl(url);

        const web = options && options.web;
        const host = options && options.host;

        // if web is true, the path resolution should be changed to reflect url resolution
        const hashUrl = url.indexOf('://') > -1 ? url.substring(url.indexOf('://') + 3, url.length) : url;
        const hash = new Buffer(hashUrl).toString('base64');

        fs.readdir(path.resolve(this.directory, hash), (error, dates) => {
            if (error) return callback(error);

            dates = dates.filter((a) => moment(a, 'Y-M-D-h:mm:ss:a')._isValid).sort((a, b) => new Date(b) - new Date(a));

            if (web) {
                return callback(null, {
                    url,
                    hash,
                    recent: `${host}/v/${hash}/${dates[0]}/index.html`,
                    dates // moment(d, 'Y-M-D-h:mm:ss:a') to be able to parse the date into a valid date
                });
            } else {
                return callback(null, {
                    url,
                    hash,
                    directory: path.resolve(this.directory, hash),
                    recent: path.resolve(this.directory, hash, dates[0], 'index.html'),
                    dates
                });
            }
        });
    }
    getAll(options, callback) {
        return fs.readdir(path.resolve(this.directory), (error, directories) => {
            async.map(directories, (value, callback) => {
                this.get(new Buffer(value, 'base64').toString('utf8'), options, callback);
            }, callback);
        });
    }
    serve(port, callback) {
        const app = express();

        app.use(compression());
        app.use(express.static('dist'));

        app.get('/api/i/*', (req, res) => {
            const site = req.params[0];

            this.get(site, {
                web: true,
                host: req.headers['host']
            }, (err, result) => {
                if (err) {
                    return res.status(500).send({
                        error: err
                    });
                } else {
                    return res.status(200).send({
                        result: result
                    });
                }
            });
        });

        app.get('/api/b/*', (req, res) => {
            const site = req.params[0];

            this.save(site, (err, result) => {
                if (err) {
                    return res.status(500).send({
                        error: err
                    });
                } else {
                    return res.status(200).send({
                        result: result
                    });
                }
            });
        });

        app.get('/api/a', (req, res) => {
            this.getAll({
                web: true,
                host: req.headers['host']
            }, (err, result) => {
                if (err) {
                    return res.status(500).send({
                        error: err
                    });
                } else {
                    return res.status(200).send({
                        result: result
                    });
                }
            });
        });

        app.get('/v/*', (req, res) => {
            const site = req.params[0];

            res.sendFile(path.resolve(this.directory, site));
        });

        app.use('*', (req, res) => {
            res.sendFile(path.resolve(__dirname, 'dist', 'index.html'));
        });

        app.listen(port, callback);
    }
}

module.exports = Timecapsule;
